### PR TITLE
Better and normalization

### PR DESCRIPTION
**Pull request by virgil-serbanuta**
This is the unoptimized version, with rather bad time complexity.

Running simplifiers without simplifier axioms seems rather dangerous.

---

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

File: kore/test/Test/Kore/Step/Simplification/And.hs [hunk 1]
```diff
@@ -406,11 +408,133 @@ test_andSimplification =
...
                     , predicate = mkNot <$> makeCeilPredicate_ fOfX
                     , substitution = mempty
                     }
         assertEqual "" mempty actual
+    , testCase "Simplifies Implies" $ do
+        let expect =
+                Conditional
+                    { term = Mock.constr10 fOfX
+                    , predicate =
+                        makeAndPredicate
```
Comment (ttuegel) [inline kore/test/Test/Kore/Step/Simplification/And.hs]: It's not very important, but I wonder: why is the `\and` left-associated here?

**Comment by ttuegel** [comment]
> Running simplifiers without simplifier axioms seems rather dangerous.

If the simplifier should be run with all the axioms, then we should not allow changing the `BuiltinAndAxiomSimplifierMap` dynamically, and this will require having a separate simplifier for the initial rule simplification (which is a good idea, anyway).

File: kore/test/Test/Kore/Step/Simplification/And.hs (hunk 1 above)
Comment (virgil-serbanuta) [reply kore/test/Test/Kore/Step/Simplification/And.hs]: I think that there is no good reason for that, the previous code used foldl1' so I used the same convention.

File: kore/src/Kore/Step/Simplification/And.hs [hunk 2]
```diff
@@ -206,15 +272,136 @@ makeEvaluateNonBool
...
+    sortedAndPredicates :: [Predicate variable]
+    sortedAndPredicates = sortByDepth andPredicates
+
+    sortByDepth :: [Predicate variable] -> [Predicate variable]
+    sortByDepth =
+        map snd
+        . sortBy (compare `on` fst)
+        . map
+            (\predicate' ->
+                ( TermLike.depth $ Predicate.unwrapPredicate predicate'
```
[the remark this answers is not in the document]
Comment (virgil-serbanuta) [reply kore/src/Kore/Step/Simplification/And.hs]: Done.

File: kore/src/Kore/Step/Simplification/And.hs [hunk 3]
```diff
@@ -206,15 +272,136 @@ makeEvaluateNonBool
...
+
+    replaceWithTopNormalized
+        :: Predicate variable
+        -> Predicate variable
+        -> Normalized (Predicate variable)
+    replaceWithTopNormalized replaceWithPredicate replaceInPredicate = do
+        let replaceIn = Predicate.unwrapPredicate replaceInPredicate
+        let replaceWith = Predicate.unwrapPredicate replaceWithPredicate
+        resultTerm <- replaceWithTop replaceWith replaceIn
+        case makePredicate resultTerm of
```
[the remark this answers is not in the document]
Comment (virgil-serbanuta) [reply kore/src/Kore/Step/Simplification/And.hs]: Done.

File: kore/src/Kore/Step/Simplification/And.hs [hunk 4]
```diff
@@ -206,15 +272,136 @@ makeEvaluateNonBool
...
+                , printError err
+                ]
+            Right p -> return p
+
+    replaceWithTop
+        :: TermLike variable
+        -> TermLike variable
+        -> Normalized (TermLike variable)
+    replaceWithTop replaceWith replaceIn
+      | replaceWith == replaceIn = Changed mkTop_
```
[the remark this answers is not in the document]
Comment (virgil-serbanuta) [reply kore/src/Kore/Step/Simplification/And.hs]: Done.

**Comment by virgil-serbanuta** [comment]
I fixed all the comments. I'll merge now and fix any further comments (if any) in a follow-up PR.

**Pull request merged**